pkg/cloud/openstack/machine/actuator.go

```
@@ -558,7 +562,136 @@ func (oc *OpenstackClient) createBootstrapToken() (string, error) {
+
+	// First, get a client for interacting with Neutron.
+	var client *gophercloud.ServiceClient
+	{
+		machineService, err := clients.NewInstanceServiceFromMachine(oc.params.KubeClient, machine)
+		if err != nil {
+			return apierrors.InvalidMachineConfiguration(fmt.Sprintf("Failed to create Neutron client: %v", err))
+		}
+
+		client = machineService.GetNetworkingClient()
+	}
+
+	// For every network specified in the Networks array of the Machine
+	// definition...
+	for _, desiredNetwork := range desiredNetworks {
+
+		// If a filter is specified, collect the matching Network
+		// UUIDs.
+		//
+		// If a filter is specified and there is no match, then return
+		// a non-nil error.
+		//
+		// If no filter is specified, let the slice empty.
+		var desiredNetworkMatchingIDs []string
+		{
+			if desiredNetwork.Filter.Name != "" || desiredNetwork.Filter.Tags != "" {
+				pager := networks.List(client, networks.ListOpts{
+					Name: desiredNetwork.Filter.Name,
+					Tags: desiredNetwork.Filter.Tags,
+				})
+
+				err := pager.EachPage(func(page pagination.Page) (bool, error) {
+					networkList, err := networks.ExtractNetworks(page)
+					if err != nil {
+						return false, err
+					}
+
+					for _, n := range networkList {
+						desiredNetworkMatchingIDs = append(desiredNetworkMatchingIDs, n.ID)
+					}
+					return true, nil
+				})
+				if err != nil {
+					return apierrors.InvalidMachineConfiguration(fmt.Sprintf("Error while reading Neutron response: %v", err))
+				}
+
+				if len(desiredNetworkMatchingIDs) == 0 {
+					return apierrors.InvalidMachineConfiguration(fmt.Sprintf("No Network found in OpenStack with Name %q and Tags %q", desiredNetwork.Filter.Name, desiredNetwork.Filter.Tags))
+				}
+			}
+		}
+
+		// for every matching NetworkID, find matching subnets
+		for _, desiredNetworkMatchingID := range desiredNetworkMatchingIDs {
```

A subnet is always attached to a network, https://docs.openstack.org/python-openstackclient/pike/cli/command-objects/subnet.html

 If a subnet exists, then it is attached to a valid network, so there is no need to validate that. You would want to just check to see that the queried subnet exists.

---

This block refers to the case in which a filter on the `network` is specified.

```
piVersion: machine.openshift.io/v1beta1
kind: Machine
metadata:
  labels:
    machine.openshift.io/cluster-api-cluster: shared-jiazha-8ck2c
    machine.openshift.io/cluster-api-machine-role: worker
    machine.openshift.io/cluster-api-machine-type: worker
    machine.openshift.io/instance-type: m1.large
    machine.openshift.io/region: regionOne
    machine.openshift.io/zone: nova
  name: shared-jiazha-8ck2c-worker-invalidsubnet
  namespace: openshift-machine-api
spec:
  metadata:
    creationTimestamp: null
  providerSpec:
    value:
      apiVersion: openstackproviderconfig.openshift.io/v1alpha1
      cloudName: openstack
      cloudsSecret:
        name: openstack-cloud-credentials
        namespace: openshift-machine-api
      flavor: m1.large
      image: rhcos-42.80.20190815.3
      kind: OpenstackProviderSpec
      metadata:
        creationTimestamp: null
      networks:
      - filter:
          name: network-filter-name1
        subnets:
        - filter:
            name: subnet-filter-name1

```

In this case, I am only looking for `{subnet.filter.name: subnet-filter-name1}` inside the context of the networks that matched `network.filter.name: network-filter-name1` (I am omitting the tags for brevity).

My assumption is: if a network filter is specified, then a subnet found in another network should not be taken into consideration for marking the subnet as valid.

WDYT?

---

The slice `desiredNetworkMatchingIDs` is only populated if a network filter has been specified. If not, the block that is used is the next one (`if len(desiredNetworkMatchingIDs)==0`)

---

Also, the filters are the same as the listopts in gophercloud for subnets and networks. I know that is badly designed and confusing, because I designed it XD. There is actually a bigger problem here though... when CAPO goes to create an instance, it will automatically validate the networks and return an error. There is no point in validating before this point because all of this is already well into runtime. I took a look at the original bugzilla, and it looks like this is really just an issue with the user being unable to read the log vomit that gets spewed out of CAPO. While it's not a bug IMO, bad logs is a big issue for CAPO.

---

> the filters are the same as the listopts in gophercloud for subnets and networks

When `List`ing subnets, I could only find an `opt` for passing `NetworkID`, and not a filter. That's why I recursively look for network and _then_ for subnets.

Am I missing something?

---

> the user being unable to read the log vomit that gets spewed out of CAPO

Yes, the entire bug is about being able to fail early and log explicitely what the issue is about. I don't think there is any functional change introduced by this early fail; only more explicit logs (and a vaguely more telling failing Event, based on the error type being specific).

You mean that this bug should not have been worked?

---

I feel like I have confused you here. Let me clarify.

## ListOpts comment
Compare https://github.com/openshift/cluster-api-provider-openstack/blob/9d6be4206d52d8849d212a76f3ef10725edc0f6a/pkg/apis/openstackproviderconfig/v1alpha1/types.go#L123

to https://github.com/gophercloud/gophercloud/blob/master/openstack/networking/v2/subnets/requests.go#L19

They are identical, and can be caste to each other. It is really confusing, and I want to change it, but its not breaking anything so its kinda on the backlog. This was just a general pointer though, nothing beyond that. 

## Error Reporting
Yes, you can validate before trying to create an instance, but the same validation runs when gophercloud tries to create the instance, and it actually does report the bug, despite it being in a horrific format. I hate to tell you that all this work was for nothing, but If I'm being honest its pretty redundant with the reporting that is already there. Check your slack, I broke down the logs there to be more readable. I think that this bug needed to be screened and vetted a little more carefully before we asked you to implement it, and that's on us. Also, I am not sure this case is even something that can happen in the installer. What you could maybe do is look at the installer code and make sure that it is impossible to create a machine object that has invalid subnets, because that would be failing fast and early.